### ISSUE TITLE
azure-cli: Could not start dynamically linked executable: /home/<user>/.azure/bin/bicep

### ISSUE THREAD

**Issue by hensou** [NixOS/nixpkgs#279870]
### Describe the bug
The `azure-cli` package includes a binary `bicep` that compiles and analyses `.bicep` files.

### Steps To Reproduce
1. Install `azure-cli`
2. Login
3. Create this simple example.bicep file:
```bicep
targetScope='subscription'

// Resource creation
resource newResourceGroup 'Microsoft.Resources/resourceGroups@2023-07-01' = {
  name: 'test'
  location: 'germanywestcentral'
  properties: {}
}

```
4. Try to deploy it with:
```console
az deployment sub create \
    --location germanywestcentral \
    --template-file example.bicep
```

### Expected behavior
It should not fail.

### Additional context
I get the following log:

```console
➜  🫠 azure-iac-experiments git:(main) ✗ az deployment sub create --location germanywestcentral                           --template-file infra/resource-group/resource-group.bicep
                                                 
Could not start dynamically linked executable: /home/hensou/.azure/bin/bicep
NixOS cannot run dynamically linked executables intended for generic
linux environments out of the box. For more information, see:
https://nix.dev/permalink/stub-ld
```

### Notify maintainers

@AkechiShiro @jonringer  

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.10, NixOS, 24.05 (Uakari), 24.05pre567011.46ae0210ce16`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `"home-manager, nixos, nixos-hardware"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by AkechiShiro** [comment]
Hi @hensou 
How did you install bicep ?

It is possible to install it by running `az bicep install` and then install it under `$HOME/.azure/bin` or `$AZURE_CONFIG_DIR` (if defined in your shell configuration file) which is clearly out of the store. Hence it's very unlikely to work as mentioned here : https://nix.dev/permalink/stub-ld

This bug I'd say is due to two current issues which are more general that we have the CLI :  
- Some configuration file of `azure-cli` is outside the nix store and the CLI can download binaries in `$HOME/.azure/bin` or in `$AZURE_CONFIG_DIR/bin` if it is defined.
- Work has started to convert the configuration file into something in the store but it's major change in the azure-cli package, see https://github.com/NixOS/nixpkgs/pull/276695 (introducing the AZURE_IMMUTABLE configuration file) 

[upstream code defines the config directory here](https://github.com/Azure/azure-cli/blob/248054ec7c0a045ff597231cd86f3b605cc79231/src/azure-cli-core/azure/cli/core/_environment.py#L10-L12)  and the [bicep subcommand seems to use it here](https://github.com/Azure/azure-cli/blob/248054ec7c0a045ff597231cd86f3b605cc79231/src/azure-cli/azure/cli/command_modules/resource/_bicep.py#L41
)

A quick workaround **if bicep was packaged in nixpkgs**, would be to configure the CLI to always use Bicep from the PATH, setting it to `bicep.use_binary_from_path=True` (output of `az bicep --help`) : 
```
az config set bicep.use_binary_from_path=True/False/if_found_in_ci
            Specify whether to use Bicep CLI from PATH or not. The default value is if_found_in_ci.
```

But that is not possible since bicep is not packaged in Nixpkgs.

**Comment by hensou** [comment]
Hi @AkechiShiro, thanks and sorry for my late reply.
I don't remember  installing `bicep` specifically, at first I thought it was installed together with the `azure-cli`, but since I also installed `vscode` and the `azure bicep extensions` there is a chance that `vscode` executed `az bicep install`.

It is interesting that from within `vscode` it does work properly, it just when I try to use it from the cli that provokes this error.

I guess for now the only option is to wait right?

**Comment by AkechiShiro** [comment]
Wait or contribute, until we can get to this issue (right now, the azure-cli fails to build in Master and tests are failing for version 2.56.0, we're focusing our efforts on fixing those issue first).

The vscode `bicep` must be statically built, if you can find the path to the file and run `file` on it or check the `ldd` output if it is not statically compiled.

**Comment by AkechiShiro** [comment]
@katexochen any input on this issue ?

**Comment by katexochen** [comment]
> A quick workaround if bicep was packaged in nixpkgs, would be to configure the CLI to always use Bicep from the PATH, setting it to bicep.use_binary_from_path=True

Yes, I think packaging bicep and consuming it from PATH is likely the best solution. @hensou could you open a separate packaging request for bicep? I haven't package a dotnet app yet. 
https://github.com/Azure/bicep

**Comment by hensou** [comment]
@katexochen, done, I created the package request in https://github.com/NixOS/nixpkgs/issues/282826 .

**Comment by katexochen** [comment]
Thanks!

> Is there anything I can do to help on https://github.com/NixOS/nixpkgs/pull/281183?

Sure! Tests currently fail, see https://github.com/NixOS/nixpkgs/pull/281183#issuecomment-1902070852 and we need to understand why.

**Comment by khaneliman** [comment]
Was creating the package in https://github.com/NixOS/nixpkgs/pull/285440 but didn't know how you wanted to join it to azure-cli.

**Comment by katexochen** [comment]
So now that we have bicep packaged, should we add it to the azure-cli closure? Would add another 150+MB. I'm not sure it's worth it. One the one hand I like if things work out of the box, but this doesn't seem something hit by many users (no thumbs up on this issue). Wdyt?

**Comment by AkechiShiro** [comment]
I think there should be a flag to override by default let's not include bicep.

**Comment by katexochen** [comment]
So `azure-cli-full` it is.. :smiling_face_with_tear: Might be an opportunity to check for other things that might not be often required, current size is just awful.

**Comment by jbiel** [comment]
I am encountering a similar issue with the `artifacttool` binary (via the `az` `azure-devops` extension.) I can open a new issue if it's more appropriate; I figured this was an OK place since this issue is still open and adjacent. I'm not sure how to get around this, but from above it seems like I can make it work by finding a static binary and setting `use_binary_from_path`. Thanks!

```
 environment.systemPackages = with pkgs; [
  (azure-cli.withExtensions [ azure-cli.extensions.azure-devops ])
];
```

```
$ az extension list
[
  {
    "experimental": false,
    "extensionType": "whl",
    "name": "azure-devops",
    "path": "/nix/store/6jy9239ddkydq9mmciyd78n61vvwd1px-azure-cli-extensions/azure-devops",
    "preview": false,
    "version": "1.0.0"
  }
]

$ az artifacts universal download ...
Downloading Universal Packages tooling (ArtifactTool_linux-x64_0.2.350): 0.80%
...
Log line: Could not start dynamically linked executable: /home/jbiel/.azure/azuredevops/cli/tools/artifacttool/ArtifactTool_linux-x64_0.2.350/artifacttool

$ ldd /home/jbiel/.azure/azuredevops/cli/tools/artifacttool/ArtifactTool_linux-x64_0.2.350/artifacttool
        linux-vdso.so.1 (0x00007fc4798f3000)
        libpthread.so.0 => /nix/store/81mi7m3k3wsiz9rrrg636sx21psj20hc-glibc-2.40-66/lib/libpthread.so.0 (0x00007fc4798c4000)
        libdl.so.2 => /nix/store/81mi7m3k3wsiz9rrrg636sx21psj20hc-glibc-2.40-66/lib/libdl.so.2 (0x00007fc4798bf000)
        libstdc++.so.6 => not found
        libm.so.6 => /nix/store/81mi7m3k3wsiz9rrrg636sx21psj20hc-glibc-2.40-66/lib/libm.so.6 (0x00007fc4797d6000)
        libgcc_s.so.1 => /nix/store/v5h01qmqjbhj3ysrd1li9364fcnlap2k-xgcc-14-20241116-libgcc/lib/libgcc_s.so.1 (0x00007fc4797a8000)
        libc.so.6 => /nix/store/81mi7m3k3wsiz9rrrg636sx21psj20hc-glibc-2.40-66/lib/libc.so.6 (0x00007fc479400000)
        /lib64/ld-linux-x86-64.so.2 => /nix/store/81mi7m3k3wsiz9rrrg636sx21psj20hc-glibc-2.40-66/lib64/ld-linux-x86-64.so.2 (0x00007fc4798f5000)
```

**Comment by naillizard** [comment]
This worked for me

shell.nix
```nix
{pkgs ? import <nixpkgs> {}}: let
  fhs = pkgs.buildFHSUserEnv {
    name = "azure-fhs";
    targetPkgs = pkgs: (with pkgs; [
      azure-cli
      openssl
      direnv
      zlib
      # Add other dependencies if needed
    ]);
  };
in
  fhs.env
```

**Comment by itpropro** [comment]
> This worked for me
> 
> shell.nix
> 
> {pkgs ? import <nixpkgs> {}}: let
>   fhs = pkgs.buildFHSUserEnv {
>     name = "azure-fhs";
>     targetPkgs = pkgs: (with pkgs; [
>       azure-cli
>       openssl
>       direnv
>       zlib
>       # Add other dependencies if needed
>     ]);
>   };
> in
>   fhs.env
> 
> then either
> 
> create an `.envrc` with the following:
> 
> ```
> use nix
> ```
> 
> or simply:
> 
> nix-shell ./shell.nix

Doesn't work anymore unfortunately, as the azure-cli build is now completely broken:

```
Error occurred loading commands!

The command failed with an unexpected error. Here is the traceback:
Attribute managed_clusters does not exist.
Traceback (most recent call last):
  File "/nix/store/gpzgfhhihpwagccjrx31llgj46s5sgsj-python3.13-azure-cli-core-2.75.0/lib/python3.13/site-packages/azure/cli/core/profiles/_shared.py", line 299, in __getattr__
    return self._operations_groups_value[item]
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^
KeyError: 'managed_clusters'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/nix/store/ai8v427i2jwj997k8v0sm72zaxc8grcr-python3.13-knack-0.12.0/lib/python3.13/site-packages/knack/cli.py", line 233, in invoke
    cmd_result = self.invocation.execute(args)
  File "/nix/store/gpzgfhhihpwagccjrx31llgj46s5sgsj-python3.13-azure-cli-core-2.75.0/lib/python3.13/site-packages/azure/cli/core/commands/__init__.py", line 666, in execute
    raise ex
  File "/nix/store/gpzgfhhihpwagccjrx31llgj46s5sgsj-python3.13-azure-cli-core-2.75.0/lib/python3.13/site-packages/azure/cli/core/commands/__init__.py", line 734, in _run_jobs_serially
    results.append(self._run_job(expanded_arg, cmd_copy))
                   ~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/gpzgfhhihpwagccjrx31llgj46s5sgsj-python3.13-azure-cli-core-2.75.0/lib/python3.13/site-packages/azure/cli/core/commands/__init__.py", line 703, in _run_job
    result = cmd_copy(params)
  File "/nix/store/gpzgfhhihpwagccjrx31llgj46s5sgsj-python3.13-azure-cli-core-2.75.0/lib/python3.13/site-packages/azure/cli/core/commands/__init__.py", line 336, in __call__
    return self.handler(*args, **kwargs)
           ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/nix/store/gpzgfhhihpwagccjrx31llgj46s5sgsj-python3.13-azure-cli-core-2.75.0/lib/python3.13/site-packages/azure/cli/core/commands/command_operation.py", line 120, in handler
    return op(**command_args)
  File "/build/azure-cli-2.75.0-src/src/azure-cli/azure/cli/command_modules/profile/custom.py", line 245, in check_cli
    raise ex
  File "/build/azure-cli-2.75.0-src/src/azure-cli/azure/cli/command_modules/profile/custom.py", line 240, in check_cli
    create_invoker_and_load_cmds_and_args(cmd.cli_ctx)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^
  File "/nix/store/gpzgfhhihpwagccjrx31llgj46s5sgsj-python3.13-azure-cli-core-2.75.0/lib/python3.13/site-packages/azure/cli/core/file_util.py", line 74, in create_invoker_and_load_cmds_and_args
    invoker.commands_loader.load_arguments()
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/nix/store/gpzgfhhihpwagccjrx31llgj46s5sgsj-python3.13-azure-cli-core-2.75.0/lib/python3.13/site-packages/azure/cli/core/__init__.py", line 548, in load_arguments
    cmd.load_arguments()  # this loads the arguments via reflection
    ~~~~~~~~~~~~~~~~~~^^
  File "/nix/store/gpzgfhhihpwagccjrx31llgj46s5sgsj-python3.13-azure-cli-core-2.75.0/lib/python3.13/site-packages/azure/cli/core/commands/__init__.py", line 321, in load_arguments
    super().load_arguments()
    ~~~~~~~~~~~~~~~~~~~~~~^^
  File "/nix/store/ai8v427i2jwj997k8v0sm72zaxc8grcr-python3.13-knack-0.12.0/lib/python3.13/site-packages/knack/commands.py", line 104, in load_arguments
    cmd_args = self.arguments_loader()
  File "/nix/store/gpzgfhhihpwagccjrx31llgj46s5sgsj-python3.13-azure-cli-core-2.75.0/lib/python3.13/site-packages/azure/cli/core/commands/command_operation.py", line 124, in arguments_loader
    op = self.get_op_handler(self.op_path)
  File "/nix/store/gpzgfhhihpwagccjrx31llgj46s5sgsj-python3.13-azure-cli-core-2.75.0/lib/python3.13/site-packages/azure/cli/core/commands/command_operation.py", line 54, in get_op_handler
    get_versioned_sdk_path(self.cli_ctx.cloud.profile, rt,
    ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
                           operation_group=self.operation_group))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/gpzgfhhihpwagccjrx31llgj46s5sgsj-python3.13-azure-cli-core-2.75.0/lib/python3.13/site-packages/azure/cli/core/profiles/_shared.py", line 503, in get_versioned_sdk_path
    api_version = getattr(api_version, operation_group)
  File "/nix/store/gpzgfhhihpwagccjrx31llgj46s5sgsj-python3.13-azure-cli-core-2.75.0/lib/python3.13/site-packages/azure/cli/core/profiles/_shared.py", line 301, in __getattr__
    raise AttributeError('Attribute {} does not exist.'.format(item))
AttributeError: Attribute managed_clusters does not exist.
```

**Comment by bearmannl** [comment]
I just started using the azure-cli / bicep in NixOS and hit this exact issue.

I can confirm that the following does work:
- Add `environment.systemPackages [ pkgs.azure-cli pkgs.bicep ];` to your nixos config/flake;
- Run `az config set bicep.use_binary_from_path=True`.
  - *Side note, it does trigger the experimental warning, `Command group 'config' is experimental and under development. Reference and support levels: https://aka.ms/CLI_refstatus`.*

As per the previous comment from ~1 year ago, the azure-cli build is not broken at this point (in the 26.05 release channel at least). So the az config unblocked me for now. Hope it helps to anyone still looking for a resolution.
